README.md
# ShareImage
![Card](https://shimg.zype.cf/v1/image?title=Generate%20Social%20Share%20Images%20Dynamically!&cloudName=zype&imagePublicID=ShareImage/Docs-Card)
## Installation
There is currently 2 Langauges Supported: **Python** and **Node.js**
And there is a **API** available to use with your project!
### API (Beta)
The API is currently in beta and may have some issues.
The domain of the API is https://shimg.zype.cf/v1
Currently, the V2 of ShareImage is only available for the Node.JS Library (with CommonJS, ESM and TypeScript support)
#### Endpoints
There is currently only one endpoint available:
##### GET `/image`
It needs Query Params in the following format:
https://shimg.zype.cf/v1/image?paramName=paramValue
The paramaters are as same as **Node.js** Params!
### Python
To Install ShareImage in Python with *PIP* Run:
```sh
pip install ShareImage
```
### Node.js
To Install ShareImage in Node.js with *NPM* Run:
```sh
npm i shareimage --save
```
Or, with *Yarn*:
```sh
yarn add shareimage
```
## Usage
Using ShareImage is a bit different across Languages.
### Python
Use the Following Code to Generate a Image and print it's URL in Python:
```py
from ShareImage import ShareImage

image = ShareImage(
    title = "My Test Image",
    cloudName = "myCloud",
    imagePublicId = "myFolder/myImage"
)

print(image)
```
### Node.js
Use the Following Code to Generate a Image and output it's URL in Node.js (ES6):
```js
import * as ShareImage from 'shareimage';

const image = await ShareImage.generateImage(
    "/path/to/image.png",
    "My awesome title",
    { type: "datauri" }
)

console.log(image)
```

## Parameters
There are only 2 Required Parameters and several optional params.
### Required Parameters
|    Parameter    |  Type  |                  Description                  |
| :-------------: | :----: | :-------------------------------------------: |
|   `src`         | String |              URL/Buffer of the Image          |
|     `title`     | String |               Title of the Image              |
### Optional Parameters
#### `output` Object (3rd parameter)
| Parameter | Type   | Default  | Description                                                                                                                                             |
| :-------: | :---:  | :-----:  | :-----------------------------------------------------------------------------------------------------------------------------------------------------: |
| `type`    | String | `base64` | The format of the output. Can be `buffer`, `datauri`, `base64`, or `file` (Save the file in FS). This option is necessary but comes with default value. |
| `options` | Object | `{}`     | This option is necessary if `type` is set to `file`. The structure of the `options` Object is `{options: {file: {path: "/file/to/output.png"}}}` |

#### `props` Object (4th parameter)
|      Parameter       |  Type   |          Default           |                            Description                             |
| :------------------: | :-----: | :------------------------: | :----------------------------------------------------------------: |
|      `tagline`       | String  |            None            |                 Tagline of Website or Tags of Post                 |
|     `titleFont`      |ShareFont|       `sirin-stencil`      |                           Font of Title                            |
|  `titleExtraConfig`  | String  |            `''`            |                         Extra Title Config                         |
| `taglineExtraConfig` | String  |            `''`            |                      Extra Config of Tagline                       |
|    `taglineFont`     |ShareFont|           arial            |                          Font of Tagline                           |
|     `imageWidth`     | Number  |            1280            |                           Width of Image                           |
|    `imageHeight`     | Number  |            669             |                          Height of Image                           |
|   `textAreaWidth`    | Number  |            760             |                         Width of TextArea                          |
|   `textLeftOffset`   | Number  |            480             |                        Left Offset of Text                         |
|    `titleGravity`    | String  |        `south_west`        |                          Gravity of Title                          |
|   `taglineGravity`   | String  |        `north_west`        |                         Gravity of Tagline                         |
|  `titleLeftOffset`   | Number  |            None            |                        Left Offset of Title                        |
| `taglineLeftOffset`  | Number  |            None            |                       Left Offset of Tagline                       |
| `titleBottomOffset`  | Number  |            254             |                       Bottom Offset of Title                       |
|  `taglineTopOffset`  | Number  |            445             |                       Top Offset of Tagline                        |
|     `textColor`      | String  |          `000000`          |                           Color of Text                            |
|     `titleColor`     | String  |            None            |  Color of Title (If not provided `textColor` will be used instead  |
|    `taglineColor`    | String  |            None            | Color of Tagline (If not provided `textColor` will be used instead |
|   `titleFontSize`    | Number  |             64             |                         Font Size of Title                         |
|  `taglineFontSize`   | Number  |             48             |                        Font Size of Tagline                        |

**Note**: If you don't provide Tagline, Parameters with `tagline` Prefix will be ignored.  
We may add more Customization to **ShareImage** in the Future.
## Sponsors
We have been sponsored by **Vercel**, **MacStadium**.  
Vercel gave us free **Pro Plan** access to host the documentation and other websites.  
MacStadium gave us free **Mac Mini Server** to host the API and for builds.  

[![Powered By Vercel](https://res.cloudinary.com/zype/image/upload/ShareImage/powered-by-vercel.png)](https://vercel.com/?utm_source=zypeoss&utm_campaign=oss)

<img src="https://res.cloudinary.com/zype/image/upload/ShareImage/MacStadium" height="44" width="212">


CHANGELOG.md
# 5.1.0

### Bug Fixes

* Fixed a bug which made it not working in ES6.

### Less Dependencies

* Replaced the existing compiler (TSC) with `tsup`, which is faster and smaller.

# 5.0.0

### Features

* Font support

### Performance Improvements

* Major Rewrite with CJS, ESM and TS support ([e805696](https://github.com/Zype-Z/ShareImage.js/commit/e805696b70820041e2c8d59b1ee9233f40f2a635))

## [1.4.6](https://github.com/Zype-Z/ShareImage.js/compare/v1.4.5...v1.4.6) (2021-09-22)


### Bug Fixes

* **config:** Fixed `release.config.js` file ([a48803e](https://github.com/Zype-Z/ShareImage.js/commit/a48803e0c3e1dd97c015ceee979266165c9170ea))

## [1.4.5](https://github.com/Zype-Z/ShareImage.js/compare/v1.4.4...v1.4.5) (2021-09-22)


### Fix

* MacStadium Banner in README [skip ci] ([66f7fb2](https://github.com/Zype-Z/ShareImage.js/commit/66f7fb22209f6a73b8020b9d8ab61d4408a80984))

## [1.4.4](https://github.com/Zype-Z/ShareImage.js/compare/v1.4.3...v1.4.4) (2021-09-13)


### Fix

* Fixed Plugin Order ([e0c8a2f](https://github.com/Zype-Z/ShareImage.js/commit/e0c8a2f4bc15fa2c555f8b3c3e5e9f97cc04e33b))

## [1.4.3](https://github.com/Zype-Z/ShareImage.js/compare/v1.4.2...v1.4.3) (2021-09-13)


### Fix

* Fixed Typos ([0448731](https://github.com/Zype-Z/ShareImage.js/commit/0448731c795dfe42497f2369b433f56d165c955a))

## [1.4.2](https://github.com/Zype-Z/ShareImage.js/compare/v1.4.1...v1.4.2) (2021-09-13)


### Fix

* Fixed Semantic Package.json ([bd9a79e](https://github.com/Zype-Z/ShareImage.js/commit/bd9a79e81d26728d9d805db48a7fdb120f5972e0))

## [1.4.1](https://github.com/Zype-Z/ShareImage.js/compare/v1.4.0...v1.4.1) (2021-09-13)


### Fix

* Added missing deps ([a0f006e](https://github.com/Zype-Z/ShareImage.js/commit/a0f006e5c0cf53613caffee3971ee1b7f31cc4c4))
* Added missing deps ([828487b](https://github.com/Zype-Z/ShareImage.js/commit/828487b1e0ad59e75e9dda650598dfcf5664bb62))
* Added missing deps ([47e13f5](https://github.com/Zype-Z/ShareImage.js/commit/47e13f5a83e6acc1f388de3acd5dcb077be5da3c))
* Fixed Semantic Release ([8204c20](https://github.com/Zype-Z/ShareImage.js/commit/8204c20becda3011e5317adddb36edd98c239316))


src/index.ts
import * as fs from "fs";
import * as path from "path";
import * as Canvas from "canvas";
import * as url from "url";

let { registerFont } = Canvas;

class ShareFont {
    [x: string]: any;
    constructor (id: string, name: string, path: fs.PathLike) {
        let options = {
            id: id,
            name: name,
            path: path,
            ShareFontOptions: {
                engine: "ShareImage.FontEngine",
                version: "v2.0.0"
            }
        }
        return options;
    }
}

type ShareFontProperties = ShareFont;

interface ShareProperties {
    tagline?: string,
    titleFont?: ShareFontProperties,
    taglineFont?: ShareFontProperties,
    titleColor?: string,
    taglineColor?: string,
    imageWidth?: number,
    imageHeight?: number,
    textAreaWidth?: number,
    textLeftOffset?: number,
    titleBottomOffset?: number,
    titleGravity?: number,
    taglineGravity?: number,
    textColor?: string,
    titleFontSize?: string,
    taglineFontSize?: string
}

interface outputOptions {
    type: "base64" | "buffer" | "datauri" | "file",
    options?: {
        file: {
            path: fs.PathLike
        }
    }
}
let sirinStencil = new ShareFont("sirin-stencil", "Sirin Stencil", path.join(__dirname, "..", "..", "assets", "fonts", "sirin-stencil.ttf"));
let arial = new ShareFont("arial", "Arial", path.join(__dirname, "..", "..", "assets", "fonts", "arial.ttf"));

let defaultOptions: ShareProperties = {
    tagline: "",
    titleFont: sirinStencil,
    taglineFont: arial,
    titleColor: "",
    imageWidth: 1280,
    imageHeight: 669,
    textLeftOffset: 480,
    titleBottomOffset: 254,
    textColor: "000000",
    titleFontSize: "64px",
    taglineFontSize: "48px"
}

interface Message {
    type: "MESSAGE" | "ERROR",
    name: string,
    message: string
}

type returnOptions = Promise<string | Buffer | undefined | Message>;

/**
 * @function
 * @name ShareImage#generateImage
 * @param {string} src - The Source file. Can be `Buffer`, `dataURI` or `URL`.
 * @param {string} title - The Title of the ShareImage.
 * @param {string} output - The output format of the Image.
 * @param {ShareProperties} props - Additional properties of the Image.
 * @return {returnOptions} Base64 URI, Data URI or Buffer. You can also save the file (returns nothing). Depends on `output` param.
 */

function generateImage(src: string | Buffer, title: string, output: outputOptions = { type: "base64" }, props: ShareProperties = defaultOptions): Promise<string | undefined | Buffer | Message> {
    let _src = src;
    let _title = title;
    let _output = output;
    let {
        tagline = "",
        titleFont = sirinStencil,
        taglineFont = arial,
        titleColor = "",
        taglineColor,
        imageWidth = 1280,
        imageHeight = 669,
        textAreaWidth,
        textLeftOffset = 480,
        titleBottomOffset = 254,
        titleGravity,
        taglineGravity,
        textColor = "000000",
        titleFontSize = "64px",
        taglineFontSize = "48px"
    } = props;

    var
        _tagline = tagline,
        _titleFont = titleFont,
        _taglineFont = taglineFont,
        _titleColor = titleColor,
        _taglineColor = taglineColor,
        _imageWidth = imageWidth,
        _imageHeight = imageHeight,
        _textAreaWidth = textAreaWidth,
        _textLeftOffset = textLeftOffset,
        _titleBottomOffset = titleBottomOffset,
        _titleGravity = titleGravity,
        _taglineGravity = taglineGravity,
        _textColor = textColor,
        _titleFontSize = titleFontSize,
        _taglineFontSize = taglineFontSize;

    let img = new Canvas.Image();
    img.src = _src;
    return new Promise<string | Buffer | undefined | Message>((resolve, reject) => {
      img.onload = () => {
        registerFont((_titleFont.path as string), { family: _titleFont.name });
        registerFont((_taglineFont.path as string), {family: _taglineFont.name});
        let canvas = Canvas.createCanvas(1280, 669);
        let ctx = canvas.getContext("2d");
        ctx.font = `${_titleFontSize} ${_titleFont.name}`
        ctx.drawImage(img, 0, 0);
        ctx.fillText(_title, _textLeftOffset, _titleBottomOffset);
        ctx.font = `${_taglineFontSize} ${_taglineFont.name}`
        ctx.fillText(_tagline, _textLeftOffset, 320);
        if (_output.type === "file") {
          let out = fs.createWriteStream(output.options.file.path);
          let stream = canvas.createPNGStream();
          stream.pipe(out);
          out.on("finish", () => {
              return;
          });
          return resolve(undefined);
        } else if (_output.type == "base64") {
          let buf: Buffer = canvas.toBuffer("image/png");
          let b64: string = buf.toString("base64");
          return resolve(b64);
        } else if (_output.type == "datauri") {
          let url: string = canvas.toDataURL();
          return resolve(url);
        } else if (_output.type == "buffer") {
          let buf: Buffer = canvas.toBuffer("image/png");
          return resolve(buf);
        } else {
          let msg: Message = {type: "ERROR", name: "UNSUPPORTED_TYPE", message: `The output type ${output.type} is not supported by ShareImage`}
          return resolve(msg);
        }
      }
   });
}

export {
    generateImage,
    ShareFont
}



src/index.mjs
/**
 * Copyright Zype Inc.
 * ===================
 * ShareImage
 * ===================
 */

/**
 * @typedef Config
 * @prop {string} title
 * @prop {string} [tagline]
 * @prop {string} cloudName
 * @prop {string} imagePublicID
 * @prop {string} [cloudinaryUrlBase]
 * @prop {string} [titleExtraConfig]
 * @prop {string} [taglineExtraConfig]
 * @prop {string} [titleFont]
 * @prop {string} [taglineFont]
 * @prop {number} [imageWidth]
 * @prop {number} [imageHeight]
 * @prop {number} [textAreaWidth]
 * @prop {number} [textLeftOffset]
 * @prop {string} [titleGravity]
 * @prop {string} [taglineGravity]
 * @prop {number} [titleLeftOffset]
 * @prop {number} [taglineLeftOffset]
 * @prop {number} [titleBottomOffset]
 * @prop {number} [taglineTopOffset]
 * @prop {string} [textColor]
 * @prop {string} [titleColor]
 * @prop {string} [taglineColor]
 * @prop {number} [titleFontSize]
 * @prop {number} [taglineFontSize]
 * @prop {string} [version]
 */

/**
 * Encodes characters for Cloudinary URL
 * Encodes some not allowed in Cloudinary parameter values twice:
 *   hash, comma, slash, question mark, backslash
 *   https://support.cloudinary.com/hc/en-us/articles/202521512-How-to-add-a-slash-character-or-any-other-special-characters-in-text-overlays-
 *
 * @param {string} text
 * @return {string}
 */
export function cleanText(text) {
    return encodeURIComponent(text).replace(/%(23|2C|2F|3F|5C)/g, '%25$1');
}

/**
 * Creates Tag String from Tags Array!
 * @param {array} tags 
 * @returns {string}
 */
export function makeTags(tags) {
    return `#${tags.join(' #')}`;
}

/**
 * Generates a social sharing image with custom text using Cloudinary’s APIs.
 *
 * @see https://cloudinary.com/documentation/image_transformations#adding_text_captions
 *
 * @param {Config} config
 * @return {string}
 */
export default function get(_a) {
    var title = _a.title,
        tagline = _a.tagline,
        cloudName = _a.cloudName,
        imagePublicID = _a.imagePublicID,
        _b = _a.cloudinaryUrlBase,
        cloudinaryUrlBase = _b === void 0 ? 'https://res.cloudinary.com' : _b,
        _c = _a.titleFont,
        titleFont = _c === void 0 ? 'futura' : _c,
        _d = _a.titleExtraConfig,
        titleExtraConfig = _d === void 0 ? '' : _d,
        _e = _a.taglineExtraConfig,
        taglineExtraConfig = _e === void 0 ? '' : _e,
        _f = _a.taglineFont,
        taglineFont = _f === void 0 ? 'arial' : _f,
        _g = _a.imageWidth,
        imageWidth = _g === void 0 ? 1280 : _g,
        _h = _a.imageHeight,
        imageHeight = _h === void 0 ? 669 : _h,
        _j = _a.textAreaWidth,
        textAreaWidth = _j === void 0 ? 760 : _j,
        _k = _a.textLeftOffset,
        textLeftOffset = _k === void 0 ? 480 : _k,
        _l = _a.titleGravity,
        titleGravity = _l === void 0 ? 'south_west' : _l,
        _m = _a.taglineGravity,
        taglineGravity = _m === void 0 ? 'north_west' : _m,
        _o = _a.titleLeftOffset,
        titleLeftOffset = _o === void 0 ? null : _o,
        _p = _a.taglineLeftOffset,
        taglineLeftOffset = _p === void 0 ? null : _p,
        _q = _a.titleBottomOffset,
        titleBottomOffset = _q === void 0 ? 254 : _q,
        _r = _a.taglineTopOffset,
        taglineTopOffset = _r === void 0 ? 445 : _r,
        _s = _a.textColor,
        textColor = _s === void 0 ? '000000' : _s,
        titleColor = _a.titleColor,
        taglineColor = _a.taglineColor,
        _t = _a.titleFontSize,
        titleFontSize = _t === void 0 ? 64 : _t,
        _u = _a.taglineFontSize,
        taglineFontSize = _u === void 0 ? 48 : _u,
        _v = _a.version,
        version = _v === void 0 ? null : _v;
    // configure social media image dimensions, quality, and format
    var imageConfig = [
        "w_" + imageWidth,
        "h_" + imageHeight,
        'c_fill',
        'q_auto',
        'f_auto',
    ].join(',');
    // configure the title text
    var titleConfig = [
        "w_" + textAreaWidth,
        'c_fit',
        "co_rgb:" + (titleColor || textColor),
        "g_" + titleGravity,
        "x_" + (titleLeftOffset || textLeftOffset),
        "y_" + titleBottomOffset,
        "l_text:" + titleFont + "_" + titleFontSize + titleExtraConfig + ":" + cleanText(title),
    ].join(',');
    // configure the tagline text
    var taglineConfig = tagline ? [
            "w_" + textAreaWidth,
            'c_fit',
            "co_rgb:" + (taglineColor || textColor),
            "g_" + taglineGravity,
            "x_" + (taglineLeftOffset || textLeftOffset),
            "y_" + taglineTopOffset,
            "l_text:" + taglineFont + "_" + taglineFontSize + taglineExtraConfig + ":" + cleanText(tagline),
        ].join(',') :
        undefined;
    // combine all the pieces required to generate a Cloudinary URL
    var urlParts = [
        cloudinaryUrlBase,
        cloudName,
        'image',
        'upload',
        imageConfig,
        titleConfig,
        taglineConfig,
        version,
        imagePublicID,
    ];
    // remove any falsy sections of the URL (e.g. an undefined version)
    var validParts = urlParts.filter(Boolean);
    // join all the parts into a valid URL to the generated image
    return validParts.join('/');
}

release.config.js
module.exports = {
    branches: [
        "+([0-9])?(.{+([0-9]),x}).x",
        "master",
        "next",
        "next-major",
        {
            name: "beta",
            prerelease: true,
        },
        {
            name: "alpha",
            prerelease: true,
        },
    ],
    plugins: [
        [
            "@semantic-release/commit-analyzer",
            {
                preset: "angular",
            },
        ],
        [
            "@semantic-release/release-notes-generator",
            {
                preset: "angular",
            },
        ],
        [
            "@semantic-release/changelog",
            {
                "changelogFile": "CHANGELOG.md"
            }
        ],
        [
          "@semantic-release/npm",
            {
               "pkgRoot": "./dist"
            }
        ],
        [
            "@semantic-release/git",
            {
                "assets": ["CHANGELOG.md", "package.json"],
                "message": "Release: ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}"
            }
        ],
        "@semantic-release/github",
    ]
};


src/index.js
"use strict";

/**
 * Copyright Zype Inc.
 * ===================
 * ShareImage
 * ===================
 */

/**
 * @typedef Config
 * @prop {string} title
 * @prop {string} [tagline]
 * @prop {string} cloudName
 * @prop {string} imagePublicID
 * @prop {string} [cloudinaryUrlBase]
 * @prop {string} [titleExtraConfig]
 * @prop {string} [taglineExtraConfig]
 * @prop {string} [titleFont]
 * @prop {string} [taglineFont]
 * @prop {number} [imageWidth]
 * @prop {number} [imageHeight]
 * @prop {number} [textAreaWidth]
 * @prop {number} [textLeftOffset]
 * @prop {string} [titleGravity]
 * @prop {string} [taglineGravity]
 * @prop {number} [titleLeftOffset]
 * @prop {number} [taglineLeftOffset]
 * @prop {number} [titleBottomOffset]
 * @prop {number} [taglineTopOffset]
 * @prop {string} [textColor]
 * @prop {string} [titleColor]
 * @prop {string} [taglineColor]
 * @prop {number} [titleFontSize]
 * @prop {number} [taglineFontSize]
 * @prop {string} [version]
 */

exports.cleanText = cleanText;
exports.makeTags = makeTags;
exports.__esModule = true;
exports.get = get;

/**
 * Encodes characters for Cloudinary URL
 * Encodes some not allowed in Cloudinary parameter values twice:
 *   hash, comma, slash, question mark, backslash
 *   https://support.cloudinary.com/hc/en-us/articles/202521512-How-to-add-a-slash-character-or-any-other-special-characters-in-text-overlays-
 *
 * @param {string} text
 * @return {string}
 */
function cleanText(text) {
    return encodeURIComponent(text).replace(/%(23|2C|2F|3F|5C)/g, '%25$1');
}

/**
 * Creates Tag String from Tags Array!
 * @param {array} tags 
 * @returns {string}
 */
function makeTags(tags) {
    return `#${tags.join(' #')}`;
}

/**
 * Generates a social sharing image with custom text using Cloudinary’s APIs.
 *
 * @see https://cloudinary.com/documentation/image_transformations#adding_text_captions
 *
 * @param {Config} config
 * @return {string}
 */
function get(_a) {
    var title = _a.title,
        tagline = _a.tagline,
        cloudName = _a.cloudName,
        imagePublicID = _a.imagePublicID,
        _b = _a.cloudinaryUrlBase,
        cloudinaryUrlBase = _b === void 0 ? 'https://res.cloudinary.com' : _b,
        _c = _a.titleFont,
        titleFont = _c === void 0 ? 'futura' : _c,
        _d = _a.titleExtraConfig,
        titleExtraConfig = _d === void 0 ? '' : _d,
        _e = _a.taglineExtraConfig,
        taglineExtraConfig = _e === void 0 ? '' : _e,
        _f = _a.taglineFont,
        taglineFont = _f === void 0 ? 'arial' : _f,
        _g = _a.imageWidth,
        imageWidth = _g === void 0 ? 1280 : _g,
        _h = _a.imageHeight,
        imageHeight = _h === void 0 ? 669 : _h,
        _j = _a.textAreaWidth,
        textAreaWidth = _j === void 0 ? 760 : _j,
        _k = _a.textLeftOffset,
        textLeftOffset = _k === void 0 ? 480 : _k,
        _l = _a.titleGravity,
        titleGravity = _l === void 0 ? 'south_west' : _l,
        _m = _a.taglineGravity,
        taglineGravity = _m === void 0 ? 'north_west' : _m,
        _o = _a.titleLeftOffset,
        titleLeftOffset = _o === void 0 ? null : _o,
        _p = _a.taglineLeftOffset,
        taglineLeftOffset = _p === void 0 ? null : _p,
        _q = _a.titleBottomOffset,
        titleBottomOffset = _q === void 0 ? 254 : _q,
        _r = _a.taglineTopOffset,
        taglineTopOffset = _r === void 0 ? 445 : _r,
        _s = _a.textColor,
        textColor = _s === void 0 ? '000000' : _s,
        titleColor = _a.titleColor,
        taglineColor = _a.taglineColor,
        _t = _a.titleFontSize,
        titleFontSize = _t === void 0 ? 64 : _t,
        _u = _a.taglineFontSize,
        taglineFontSize = _u === void 0 ? 48 : _u,
        _v = _a.version,
        version = _v === void 0 ? null : _v;
    // configure social media image dimensions, quality, and format
    var imageConfig = [
        "w_" + imageWidth,
        "h_" + imageHeight,
        'c_fill',
        'q_auto',
        'f_auto',
    ].join(',');
    // configure the title text
    var titleConfig = [
        "w_" + textAreaWidth,
        'c_fit',
        "co_rgb:" + (titleColor || textColor),
        "g_" + titleGravity,
        "x_" + (titleLeftOffset || textLeftOffset),
        "y_" + titleBottomOffset,
        "l_text:" + titleFont + "_" + titleFontSize + titleExtraConfig + ":" + cleanText(title),
    ].join(',');
    // configure the tagline text
    var taglineConfig = tagline ? [
            "w_" + textAreaWidth,
            'c_fit',
            "co_rgb:" + (taglineColor || textColor),
            "g_" + taglineGravity,
            "x_" + (taglineLeftOffset || textLeftOffset),
            "y_" + taglineTopOffset,
            "l_text:" + taglineFont + "_" + taglineFontSize + taglineExtraConfig + ":" + cleanText(tagline),
        ].join(',') :
        undefined;
    // combine all the pieces required to generate a Cloudinary URL
    var urlParts = [
        cloudinaryUrlBase,
        cloudName,
        'image',
        'upload',
        imageConfig,
        titleConfig,
        taglineConfig,
        version,
        imagePublicID,
    ];
    // remove any falsy sections of the URL (e.g. an undefined version)
    var validParts = urlParts.filter(Boolean);
    // join all the parts into a valid URL to the generated image
    return validParts.join('/');
}
exports["default"] = get;